drivers/platform/aducm3029/uart.c

```diff
@@ -0,0 +1,401 @@
+
+/**
+ * @brief Allocates the memory needed for the UART descriptor
+ * @return Address to the allocated memory, NULL if the allocation fails.
+ */
+static struct uart_desc *alloc_desc_mem(void)
+{
+	struct uart_desc	*desc;
+	struct aducm_uart_desc	*aducm_desc;
+	uint32_t		mem;
+
+	desc = calloc(1, sizeof(struct uart_desc));
+	if (!desc)
+		return NULL;
+	aducm_desc = calloc(1, sizeof(struct aducm_uart_desc));
+	if (!aducm_desc) {
+		free(desc);
+		return NULL;
+	}
+	desc->extra = aducm_desc;
+
+	/* Allocate and align buffer to 32 bits */
+	aducm_desc->adi_uart_buffer = calloc(1, ADI_UART_BIDIR_MEMORY_SIZE + 3);
+	if (!aducm_desc->adi_uart_buffer) {
+		free(aducm_desc);
+		free(desc);
+		return NULL;
+	}
+
+	mem = (uint32_t)aducm_desc->adi_uart_buffer;
+	aducm_desc->adi_uart_buffer = (uint8_t *)((mem+3u) & (~(3u)));
+	aducm_desc->adi_uart_buffer_offset =
+		(uint32_t)aducm_desc->adi_uart_buffer - mem;
+
+	return desc;
+}
+
+/**
+ * @brief Deallocates the memory needed for the UART descriptor
+ * @param desc - Descriptor to be deallocated
+ */
+static void free_desc_mem(struct uart_desc *desc)
+{
+	struct aducm_uart_desc	*aducm_desc = desc->extra;
+
+	free((void *)((uint32_t)aducm_desc->adi_uart_buffer -
+		      aducm_desc->adi_uart_buffer_offset));
+	aducm_desc->adi_uart_buffer = NULL;
+	aducm_desc->adi_uart_buffer_offset = 0;
+	free(desc->extra);
+	free(desc);
+}
+
+/**
+ * @brief Call the user defined callback when a read/write operation completed
+ * @param desc:		Descriptor of the UART device
+ * @param event:	Event ID from ADI_UART_EVENT
+ * @param buff:		Pointer to the handled buffer or to an error code
+ */
+static void uart_callback(void *desc, uint32_t event, void *buff)
+{
+	struct aducm_uart_desc *extra = ((struct uart_desc *)desc)->extra;
+
+	switch(event) {
+	case ADI_UART_EVENT_RX_BUFFER_PROCESSED:
+		extra->waiting_read_callback--;
+		extra->callback(extra->param, READ_DONE, buff);
+		break;
+	case ADI_UART_EVENT_TX_BUFFER_PROCESSED:
+		extra->waiting_write_callback--;
+		extra->callback(extra->param, WRITE_DONE, buff);
+		break;
+	default:
+		extra->errors |= (uint32_t)buff;
+		extra->callback(extra->param, ERROR, buff);
+		break;
+	}
+}
+
+/**
+ * @brief Calls the low level read functions depending on the mode
+ * @param desc:	Descriptor of the UART device
+ * @param data:	Buffer where data will be read
+ * @param bytes_number:	Number of bytes to be read. Between 1 and 1024
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t uart_read(struct uart_desc *desc, uint8_t *data,
+		  uint32_t bytes_number)
+{
+	uint32_t errors;
+	struct aducm_uart_desc *extra;
+	extra = desc->extra;
+
+	if (bytes_number == 0 || bytes_number > MAX_BYTES) {
+		errors = BAD_INPUT_PARAMETERS;
+		goto failure;
+	}
+
+	if (extra->mode == BLOCKING_MODE) {
+		ADI_UART_RESULT ret;
+
+		if (extra->waiting_read_callback) {
+			errors = WAITING_FOR_CALLBACK;
+			goto failure;
+		}
+		ret = adi_uart_Read(
```

Comment:
when calling a function start from the first line with the parameters don't lave it empty.